### PR TITLE
Add ArrayIndentationFixer

### LINKED ISSUE #770: Wrapped array elements indentation not fixed.
Hi, 

I came across an issue where elements in array that are wrapped with double indentation, which isn't fixed by the fixer.

```
<?php
// expected
$var = array(
    'a' => 1,
    'b' => 2,
);
// actual
$var = array(
        'a' => 1,
        'b' => 2,
);
```

I've tried adding this case to the indentation fixer but it does not fail on unit test.

Can someone please verify or let me know it's not an issue?

Thanks.

### PR DESCRIPTION
**Pull request by julienfalque**
Closes #770 and #1438.

### CODE REVIEW

File: src/Fixer/Whitespace/ArrayIndentationFixer.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    private function findArrays(Tokens $tokens)
+    {
+        $arrays = [];
+
+        foreach ($this->findArrayTokenRanges($tokens, 0, count($tokens) - 1) as $arrayTokenRanges) {
+            $arrays[] = [
+                'start' => $arrayTokenRanges[0][0],
+                'end' => $arrayTokenRanges[count($arrayTokenRanges) - 1][1],
+                'token_ranges' => $arrayTokenRanges,
+            ];
```
Comment (SpacePossum) [inline src/Fixer/Whitespace/ArrayIndentationFixer.php]: maybe
```php

// remove $arrays

foreach(...) {
    $array = [
        'start' => $arrayTokenRanges[0][0],
        'end' => $arrayTokenRanges[count($arrayTokenRanges) - 1][1],
        'token_ranges' => $arrayTokenRanges,
    ];
    $array['line_braces_deltas'] = $this->computeArrayLineBracesDeltas($tokens, $array);

    yield $array;
}

// remove next foreach and return statement
```

Comment (julienfalque) [reply src/Fixer/Whitespace/ArrayIndentationFixer.php]: I removed the second `foreach` but didn't use `yield`.

**Comment by EikeDawid** [comment]
@julienfalque are you still working on this PR or has it been abandoned? - I would love to have a consistent way to format arrays.

**Comment by julienfalque** [comment]
The PR is not abandoned but it needs more review. If you have time, any help is welcome :)

**Comment by keradus** [comment]
@EikeDawid, why won't you help and make a decent review? it would help merging the PR for sure

**Comment by EikeDawid** [comment]
@keradus let me see if i can find somebody

**Comment by mgarciadelojo** [comment]
Hi!

First of all, thank you @julienfalque for the Pull Request. Looks amazing!

I've been playing with it for a while, and everything seems fine to me. You can see a few examples below:

**Input:**

```
<?php

$array = array(
		'a' => 1,
		'asdf'      => 2,
		'adddd' => 3,
		'c' => 4,
	);


$array = array(
		 1,
		2,
		  3,
		4,
		    5,
		6
		);


$array = array(1,2,3,4,5,    );


$this->assertEquals(
    [
    'firstKey' 		=> '50',
    'secondKey' 		=> '90',
    'whateverKey' 		=> '/tmp/report',
    ],
    $this->configuration->getLoggingConfiguration()
);

return array(
  'first-name'	 => 'First Name',
  'last-name' 	 => 'Last Name',
  'language' => array(
    	'en' => 'English',
    	'es'  =>    'Spanish',
  ),
);
```

**Output:**

```
<?php

$array = [
    'a' => 1,
    'asdf' => 2,
    'adddd' => 3,
    'c' => 4,
];

$array = [
    1,
    2,
    3,
    4,
    5,
    6,
];

$array = [1, 2, 3, 4, 5];

$this->assertSame(
    [
        'firstKey' => '50',
        'secondKey' => '90',
        'whateverKey' => '/tmp/report',
    ],
    $this->configuration->getLoggingConfiguration()
);

return [
    'first-name' => 'First Name',
    'last-name' => 'Last Name',
    'language' => [
        'en' => 'English',
        'es' => 'Spanish',
    ],
];
```

The only thing I am missing are more tests, but apart from that, works great.

Cheers

File: README.rst [hunk 2]
```diff
@@ -247,6 +247,10 @@ Choose from the list of available rules:
     whose lines all start with an asterisk (``phpdocs_like``) or any
     multi-line comment (``all_multiline``); defaults to ``'phpdocs_only'``
 
+* **array_indentation**
+
+  Multiline arrays must be indented with configured indentation.
```
Comment (keradus) [inline README.rst]: actually, I would drop ` with configured indentation` part, it's misleading, we don't have config for this fixer.

you meant global config, but that is used across many rules, and we don't mention that "global config" for them.

File: src/Fixer/Whitespace/ArrayIndentationFixer.php [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+                    $content = preg_replace(
+                        '/(\R)'.preg_quote($previousLineInitialIndent, '/').'([\t ]*)$/',
+                        '$1'.$previousLineNewIndent.'$2',
+                        $token->getContent()
+                    );
+                }
+
+                $previousLineBracesDelta = $lineBracesDelta;
+
+                $tokens[$index] = new Token([T_WHITESPACE, $content]);
```
Comment (keradus) [inline src/Fixer/Whitespace/ArrayIndentationFixer.php]: just double-checking, does this fixer any time inserting new token to the collection?

File: tests/Fixer/Whitespace/ArrayIndentationFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+                ,
+                <<<'INPUT'
+<?php
+$foo = [
+  'foo',
+        'bar' => 'baz',
+ ];
+INPUT
+                ,
+            ],
```
Comment (keradus) [inline tests/Fixer/Whitespace/ArrayIndentationFixerTest.php]: it shall pass, but please add
```php
            [
                <<<'EXPECTED'
<?php
$foo = [
    'foo', // comment
    'bar',
];
EXPECTED
                ,
                <<<'INPUT'
<?php
$foo = [
  'foo', // comment
'bar',
 ];
INPUT
                ,
            ],
```

just to be sure

File: README.rst (hunk 2 above)
Comment (julienfalque) [reply README.rst]: Updated.

File: src/Fixer/Whitespace/ArrayIndentationFixer.php (hunk 3 above)
Comment (julienfalque) [reply src/Fixer/Whitespace/ArrayIndentationFixer.php]: It doesn't.

File: tests/Fixer/Whitespace/ArrayIndentationFixerTest.php (hunk 4 above)
Comment (julienfalque) [reply tests/Fixer/Whitespace/ArrayIndentationFixerTest.php]: Added.

**Comment by julienfalque** [comment]
WIP as I found test cases that currently fail.

**Comment by julienfalque** [comment]
I refactored the fixer a bit, added and fixed test cases. I feel more confident about the stability of the fixer now. I found an edge case that is still not supported though:

```diff
 $foo = [
     0 =>
-        'bar',
+    'bar',
     1 =>
-        'baz',
+    'baz',
 ];
```
IMO this case should be left untouched, i.e. values that are on their own line (rather than after `=>`) should be indented once. But I assume this is a very rare edge case (found only once in Symfony codebase), I'd say this should be a separate PR so this one can get merged. WDYT?

**Comment by keradus** [comment]
I would say that for this case we do have `no_multiline_whitespace_around_double_arrow` rule ;)

**Comment by keradus** [comment]
Thank you @julienfalque.

**Comment by keradus** [comment]
really nice here !

**Comment by julienfalque** [comment]
Really happy to see this merged, thanks! :)

**Comment by mgarciadelojo** [comment]
Amazing, thanks guys

**Pull request merged**